proposals/0048-container-api.md

```diff
@@ -0,0 +1,193 @@
+- `AstroContainer.renderToString`: renders a component and return a string.
+- `AstroContainer.renderToResponse`: renders a component and returns the `Response` emitted by the rendering phase.
+
+### `create` function
+
+A container is class exposed via the `astro/container` specifier. Users **shouldn't** create a container using the `new` instance, but they should use the static function `create`:
+
+```js
+const container = await AstroContainer.create()
+```
+
+The reason why `create` returns a promise is because internally we will validate and parse the user configuration. As a developer, you will be able to pass the same configuration that Astro uses:
+
+```js
+import astroConfig from "../src/astro.config.mjs";
+
+const container = await AstroContainer.create({
+    astroConfig
+})
+```
+
+The function `create` will throw an error if there's a validation error.
+
+> [!NOTE]
+> If you use a TypeScript file for your configuration, you are responsible for loading and transforming it to JavaScript.
+
+
+The `container` binding will expose a `renderToString` that accepts a Astro component and will return a `string`:
+
+```js
+import astroConfig from "../src/astro.config.mjs";
+
+const container = await AstroContainer.create({
+    astroConfig
+})
+
+const content = await container.renderToString(AstroComponent);
+```
+
+#### Options
+
+It will be possible to tweak the container APIs with options and such. The `.create` function will accept the following interface:
+
+```ts
+type AstroContainerOptions = {
+    streaming: boolean;
+    renderers: AstroRenderer[];
+    astroConfig: AstroUserConfig;
+}
+```
+
+The `astroConfig` object is literally the same object exposed by the `defineConfig`, inside the `astro.config.mjs` file. This very configuration 
+will go under the same schema validation that Astro uses internally. This means that an invalid schema will result in an error of the `create` function.
+
+
+### `renderToString` and `renderToResponse` options
+
+This function can accept an optional object as a second parameter:
+
+```js
+import { onRequest } from "../src/middleware.js"
+
+const response = await container.renderToString(Card, {
+    slots: {
+        default: await container.renderToString(CardItem) 
+    },
+    request: new Request("https://example.com/blog/blog-slug", {
+        headers: {}
+    }),
+    params: {
+        "slug": "blog-slug"
+    }, // in case your route is `pages/blog/[slug].astro`
+    locals: {
+        someFn() {
+            
+        },
+        someString: "string",
+        someNumber: 100
+    },
+});
+```
+
+- `slots`: required in case your component is designed to render some slots inside of it. It supposed named slots too.
+- `request`: required in case your component/page access to some information such as `Astro.url` or `Astro.request`.
+- `params`: the `Astro.params` to provide to the components 
+- `locals`: initial value of the `Astro.locals`.
+- `routeType`: useful to test endpoints.
```

Comment:
Seems it lacks `props`. Can I know why they are not here?

Reply:
I update this document every time things change. When I released the container API, the props weren't there. 

The latest patch allows you to do that now. I will update the document soon

Reply:
Great! :+1:

I'm looking forward to the update.

Reply:
It was released this morning!